src/hash.rs

```diff
@@ -64,3 +70,65 @@ fn file_hash(algorithm: &str, path: &str) -> Result<String> {
 
     Ok(hash_algorithm(algorithm, &bytes)?)
 }
+
+fn totp_generate(hex_seed: &str) -> Result<String> {
+    let mut seed: [u8; 64] = [0; 64];
+
+    let hex_seed_bytes = (0..hex_seed.len())
+                        .step_by(2)
+                        .map(|i| u8::from_str_radix(&hex_seed[i..i + 2], 16));
+
+    let mut i = 0;
+    for b in hex_seed_bytes {
+        seed[i] = b.unwrap();
+        i += 1;
+    }
+
+    let mut ipad: [u8; 64] = [0; 64];
+    let mut opad: [u8; 64] = [0; 64];
+
+    for j in 0..64 {
+        ipad[j] = seed[j] ^ 0x36;
+        opad[j] = seed[j] ^ 0x5C;
+    }
+    let time: u64 = SystemTime::now().duration_since(UNIX_EPOCH).expect("SystemTime before UNIX EPOC").as_secs() / 30;
+
+    let time_bytes: [u8; 8] = time.to_be_bytes();
+
+    let mut ipad_time: [u8; 72] = [0; 72];
+
+    for j in 0..72 {
+        if j < 64 {
+            ipad_time[j] = ipad[j];
+        } else {
+            ipad_time[j] = time_bytes[j - 64];
+        }
+    }
```

Comment:
You _could_ get rid of this loop by using `slice::clone_from_slice` like so:

```rust
    let mut ipad_time: [u8; 72] = [0; 72];
    ipad_time[..64].clone_from_slice(&ipad[..]);
    ipad_time[64..].clone_from_slice(&time_bytes[..]);
```

**But**, I think this ipad_time array (and hence the copy operations) shouldn't be necessary at all. You can pass the data into the Sha1 hasher using multiple calls to `Sha1::update`.

Here's an example of that in the Rust playground. It's using Sha2 because that's all that is available there: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4ab98d46d41b57af58e2d6c740936149